Modix.Bot/Behaviors/PromotionPollBehavior/PromotionDialogBehavior.cs

```diff
@@ -0,0 +1,211 @@
+            IMemoryCache memoryCache,
+            IPromotionsService promotionsService,
+            IImageService imageService,
+            IAuthorizationService authorizationService,
+            IPromotionCampaignRepository promotionCampaignRepository,
+            IPromotionDialogRepository promotionDialogRepository
+            )
+        {
+            _designatedChannelService = designatedChannelService;
+            _discordSocketClient = discordSocketClient;
+            _memoryCache = memoryCache;
+            _promotionsService = promotionsService;
+            _imageService = imageService;
+            _promotionCampaignRepository = promotionCampaignRepository;
+            _authorizationService = authorizationService;
+            _promotionDialogRepository = promotionDialogRepository;
+        }
+
+        public async Task HandleNotificationAsync(PromotionActionCreatedNotification notification,
+            CancellationToken cancellationToken = default)
+        {
+            if (cancellationToken.IsCancellationRequested)
+                return;
+
+            switch (notification.Data.Type)
+            {
+                case PromotionActionType.CampaignCreated:
+                    await CreatePromoDialog(notification.Data);
+                    break;
+                case PromotionActionType.CampaignClosed:
+                    await DeletePromoDialog(notification.Data);
+                    break;
+            }
+        }
+
+        public async Task HandleNotificationAsync(ReactionAddedNotification notification,
+            CancellationToken cancellationToken = default)
+        {
+            var reaction = notification.Reaction.Emote;
+            var user = (IGuildUser)notification.Reaction.User.Value;
+            var message = await notification.Message.GetOrDownloadAsync();
+
+            if (cancellationToken.IsCancellationRequested
+                || notification.Reaction.User.Value.IsBot
+                || message.Embeds.Count < 1)
+                return;
+
+            if (_memoryCache.TryGetValue(message.Id, out CachedPromoDialog poll)
+            && ValidateVoteReaction(reaction))
+            {
+                string voteError = null;
+                var sentiment = reaction.Name == _approveEmoji.Name ? PromotionSentiment.Approve : PromotionSentiment.Oppose;
+
+                await message.RemoveReactionAsync(notification.Reaction.Emote, notification.Reaction.User.Value);
+
+                try
+                {
+                    await _authorizationService.OnAuthenticatedAsync(user);
+                    await _promotionsService.UpdateOrAddComment(poll.CampaignId, sentiment, "This vote was cast from Discord");
+                }
+                catch (Exception e)
+                {
+                    voteError = e.Message;
+                }
+
+                var campaign = await _promotionCampaignRepository.GetCampignSummaryByIdAsync(poll.CampaignId);
```

Comment:
Would it not make sense to put this also up inside the `try`? If there was an error with the vote, the campaign shouldn't have changed, in theory, yes?

Reply:
if that call throws then the embed cant be built anyway because we wont have the vote counts. so i suppose it needs to be checked for null before moving on to the embeds